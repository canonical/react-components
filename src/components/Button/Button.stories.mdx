import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import Button from "./Button";

<Meta
  title="Button"
  component={Button}
  argTypes={{
    children: {
      control: {
        type: "text",
      },
      defaultValue: "Click me!",
    },
  }}
/>

export const Template = (args) => <Button {...args} />;

### Button

This is a [React](https://reactjs.org/) component for the Vanilla [Button](https://docs.vanillaframework.io/patterns/buttons/).

Buttons are clickable elements used to perform an action, they can be used for buttons and link elements.

### Props

<ArgsTable of={Button} />

### Default

<Canvas>
  <Story
    name="Default"
    args={{
      children: "Button",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<Canvas>
  <Story
    name="Default disabled"
    args={{
      children: "Button",
      disabled: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Base

<Canvas>
  <Story name="Base">
    <Button appearance="base">Base button</Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Base disabled">
    <Button appearance="base" disabled>
      Base button disabled
    </Button>
  </Story>
</Canvas>

### Link

<Canvas>
  <Story name="Link">
    <Button element="a" href="#test" appearance="base">
      Link button
    </Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Link disabled">
    <Button element="a" href="#test" appearance="base" disabled>
      Link button disabled
    </Button>
  </Story>
</Canvas>

### Positive

<Canvas>
  <Story name="Positive">
    <Button appearance="positive">Positive button</Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Positive disabled">
    <Button appearance="positive" disabled>
      Positive button disabled
    </Button>
  </Story>
</Canvas>

### Negative

<Canvas>
  <Story name="Negative">
    <Button appearance="negative">Negative button</Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Negative disabled">
    <Button appearance="negative" disabled>
      Negative button disabled
    </Button>
  </Story>
</Canvas>

### Brand

<Canvas>
  <Story name="Brand">
    <Button appearance="brand">Brand button</Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Brand disabled">
    <Button appearance="brand" disabled>
      Brand button disabled
    </Button>
  </Story>
</Canvas>

### Inline

<Canvas>
  <Story name="Inline">
    <>
      <span>Everything you need to get started with Vanilla. </span>
      <Button appearance="neutral" inline>
        Inline button
      </Button>
    </>
  </Story>
</Canvas>

### Dense

<Canvas>
  <Story name="Dense">
    <>
      <span>Everything you need to get started with Vanilla. </span>
      <Button dense>Dense button</Button>
    </>
  </Story>
</Canvas>

### Small

<Canvas>
  <Story name="Small">
    <>
      <Button small>Small button</Button>
      <Button small dense>
        Small dense button
      </Button>
    </>
  </Story>
</Canvas>

### Icon

<Canvas>
  <Story name="Icon">
    <Button hasIcon>
      <i className="p-icon--plus"></i>
    </Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Icon & text">
    <Button hasIcon>
      <i className="p-icon--plus"></i> <span>Button with icon & text</span>
    </Button>
  </Story>
</Canvas>
