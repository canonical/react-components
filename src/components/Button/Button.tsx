import classNames from "classnames";
import React from "react";
import type {
  ButtonHTMLAttributes,
  ComponentType,
  ElementType,
  MouseEventHandler,
  ReactNode,
} from "react";

import type { ClassName, ValueOf } from "types";

export const ButtonAppearance = {
  BASE: "base",
  BRAND: "brand",
  DEFAULT: "",
  LINK: "link",
  NEGATIVE: "negative",
  POSITIVE: "positive",
} as const;

/**
 * The type of the Button props.
 * @template P - The type of the props if providing a component to `element`
 */
export type Props<P = null> = {
  /**
   * The appearance of the button.
   */
  appearance?: ValueOf<typeof ButtonAppearance> | string;
  /**
   * The content of the button.
   */
  children?: ReactNode;
  /**
   * Optional class(es) to pass to the button element.
   */
  className?: ClassName;
  /**
   * Whether the button should have dense padding.
   */
  dense?: boolean;
  /**
   * Whether the button should be disabled.
   */
  disabled?: boolean;
  /**
   * Optional element or component to use instead of `button`.
   */
  element?: ElementType | ComponentType<P>;
  /**
   * Whether the button has an icon in the content.
   */
  hasIcon?: boolean;
  /**
   * Whether the button should display inline.
   */
  inline?: boolean;
  /**
   * Function for handling button click event.
   */
  onClick?: MouseEventHandler<HTMLButtonElement>;
  /**
   * Whether the button should be small.
   */
  small?: boolean;
} & (Omit<ButtonHTMLAttributes<HTMLButtonElement>, "onClick"> | P);

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Button](https://docs.vanillaframework.io/patterns/buttons/).
 *
 * Buttons are clickable elements used to perform an action, they can be used for buttons and link elements.
 * @template P - The type of the props if providing a component to `element`
 */
const Button = <P,>({
  appearance,
  children,
  className,
  dense,
  disabled,
  element: Component = "button",
  hasIcon,
  inline,
  onClick,
  small,
  ...buttonProps
}: Props<P>): JSX.Element => {
  const classes = classNames(
    appearance ? `p-button--${appearance}` : "p-button",
    {
      "has-icon": hasIcon,
      "is-dense": dense,
      "is-disabled": disabled,
      "is-inline": inline,
      "is-small": small,
    },
    className
  );
  const onClickDisabled = (e: MouseEvent) => e.preventDefault();

  return (
    <Component
      className={classes}
      onClick={disabled ? onClickDisabled : onClick}
      aria-disabled={disabled || undefined}
      {...buttonProps}
    >
      {children}
    </Component>
  );
};

export default Button;
