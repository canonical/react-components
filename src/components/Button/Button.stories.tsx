import React, { HTMLProps } from "react";
import { Meta, StoryObj } from "@storybook/react";

import Tooltip from "components/Tooltip";

import Button, { ButtonAppearance } from "./Button";

const meta: Meta<typeof Button> = {
  component: Button,
  tags: ["autodocs"],

  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },

    appearance: {
      control: {
        options: ButtonAppearance,
        type: "select",
      },
    },
  },

  args: {
    children: "Click me!",
    appearance: ButtonAppearance.DEFAULT,
  },
};

export default meta;

type Story<P = null> = StoryObj<typeof Button<P>>;

export const Default: Story = {
  name: "Default",

  args: {
    children: "Button",
  },
};

export const DefaultDisabled: Story = {
  name: "Default disabled",

  args: {
    children: "Button",
    disabled: true,
  },
};

export const Base: Story = {
  name: "Base",

  args: {
    children: "Base button",
    appearance: "base",
  },
};

export const BaseDisabled: Story = {
  name: "Base disabled",

  args: {
    children: "Base button disabled",
    appearance: "base",
    disabled: true,
  },
};

export const Link = {
  name: "Link",

  args: {
    children: "Link button",
    appearance: "base",
    element: "a",
    href: "#test",
  },
};

export const LinkDisabled: Story<HTMLProps<HTMLAnchorElement>> = {
  name: "Link disabled",

  args: {
    children: "Link button disabled",
    appearance: "base",
    element: "a",
    href: "#test",
    disabled: true,
  },
};

export const Positive: Story = {
  name: "Positive",

  args: {
    children: "Positive button",
    appearance: "positive",
  },
};

export const PositiveDisabled: Story = {
  name: "Positive disabled",

  args: {
    children: "Positive button disabled",
    appearance: "positive",
    disabled: true,
  },
};

export const Negative: Story = {
  name: "Negative",

  args: {
    children: "Negative button",
    appearance: "negative",
  },
};

export const NegativeDisabled: Story = {
  name: "Negative disabled",

  args: {
    children: "Negative button disabled",
    appearance: "negative",
    disabled: true,
  },
};

export const Brand: Story = {
  name: "Brand",

  args: {
    children: "Brand button",
    appearance: "brand",
  },
};

export const BrandDisabled: Story = {
  name: "Brand disabled",

  args: {
    children: "Brand button disabled",
    appearance: "brand",
    disabled: true,
  },
};

export const Inline: Story = {
  render: () => (
    <>
      <span>Everything you need to get started with Vanilla.</span>
      <Button appearance="neutral" inline>
        Inline button
      </Button>
    </>
  ),

  name: "Inline",
};

export const Dense: Story = {
  render: () => (
    <>
      <span>Everything you need to get started with Vanilla.</span>
      <Button dense>Dense button</Button>
    </>
  ),

  name: "Dense",
};

export const Small: Story = {
  render: () => (
    <>
      <Button small>Small button</Button>
      <Button small dense>
        Small dense button
      </Button>
    </>
  ),

  name: "Small",
};

export const Icon: Story = {
  name: "Icon",

  args: {
    children: <i className="p-icon--plus"></i>,
    hasIcon: true,
  },
};

export const IconText: Story = {
  name: "Icon & text",

  args: {
    children: (
      <>
        <i className="p-icon--plus"></i> <span>Button with icon & text</span>
      </>
    ),

    hasIcon: true,
  },
};

export const IconWithTooltip: Story = {
  name: "Disabled with tooltip",

  args: {
    children: "Disabled button with a tooltip",
    disabled: true,
  },
  render: (args) => (
    <div style={{ paddingTop: "3rem" }}>
      <Tooltip message="This button is disabled">
        <Button {...args} />
      </Tooltip>
    </div>
  ),
};
