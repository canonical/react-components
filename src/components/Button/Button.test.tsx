import { shallow } from "enzyme";
import React from "react";

import Button from "./Button";

describe("Button ", () => {
  it("renders", () => {
    const wrapper = shallow(<Button>Test content</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders as a link", () => {
    const wrapper = shallow(<Button element="a">Test content</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it("can handle button click events", () => {
    const onClick = jest.fn();
    const wrapper = shallow(<Button onClick={onClick} />);
    wrapper.simulate("click");
    expect(onClick).toHaveBeenCalled();
  });

  it("can handle anchor click events", () => {
    const onClick = jest.fn();
    const wrapper = shallow(<Button element="a" onClick={onClick} />);
    wrapper.simulate("click");
    expect(onClick).toHaveBeenCalled();
  });

  it("correctly disables a button", () => {
    const onClick = jest.fn();
    const wrapper = shallow(<Button disabled={true} onClick={onClick} />);
    expect(wrapper.prop("disabled")).toBe(true);
    expect(wrapper.prop("className").includes("is-disabled")).toBe(false);
    expect(wrapper.prop("aria-disabled")).toBe(undefined);
    const preventDefault = jest.fn();
    wrapper.simulate("click", { preventDefault });
    expect(preventDefault).toHaveBeenCalled();
    expect(onClick).not.toHaveBeenCalled();
  });

  it("correctly disables a link", () => {
    const onClick = jest.fn();
    const wrapper = shallow(
      <Button element="a" disabled={true} onClick={onClick} />
    );
    expect(wrapper.prop("className").includes("is-disabled")).toBe(true);
    expect(wrapper.prop("aria-disabled")).toBe(true);
    expect(wrapper.prop("disabled")).toBe(undefined);
    const preventDefault = jest.fn();
    wrapper.simulate("click", { preventDefault });
    expect(preventDefault).toHaveBeenCalled();
    expect(onClick).not.toHaveBeenCalled();
  });

  it("correctly handle icons", () => {
    const wrapper = shallow(<Button hasIcon={true} />);
    expect(wrapper.prop("className").includes("has-icon")).toBe(true);
  });

  it("can be inline", () => {
    const wrapper = shallow(<Button inline={true} />);
    expect(wrapper.prop("className").includes("is-inline")).toBe(true);
  });

  it("can be dense", () => {
    const wrapper = shallow(<Button dense />);
    expect(wrapper.prop("className").includes("is-dense")).toBe(true);
  });

  it("can be small", () => {
    const wrapper = shallow(<Button small />);
    expect(wrapper.prop("className").includes("is-small")).toBe(true);
  });

  it("can add additional classes", () => {
    const wrapper = shallow(<Button className="extra-class" />);
    const className = wrapper.prop("className");
    expect(className.includes("p-button")).toBe(true);
    expect(className.includes("extra-class")).toBe(true);
  });

  it("puts additional classes at the end", () => {
    const wrapper = shallow(<Button className="extra-class" dense />);
    expect(wrapper.prop("className")).toEqual("p-button is-dense extra-class");
  });

  it("handles base button props", () => {
    const wrapper = shallow(<Button type="button" />);
    wrapper.simulate("click");
    expect(wrapper.prop("type")).toBe("button");
  });

  it("handles alternate element props", () => {
    const wrapper = shallow(<Button element="a" href="http://example.com" />);
    wrapper.simulate("click");
    expect(wrapper.prop("href")).toBe("http://example.com");
  });

  it("handles supplied component props", () => {
    type LinkProps = {
      /** An address */
      to: string;
      children: React.ReactNode;
    };
    const Link = ({ children, to }: LinkProps) => <a href={to}>{children}</a>;
    const wrapper = shallow(<Button element={Link} to="http://example.com" />);
    wrapper.simulate("click");
    expect(wrapper.prop("to")).toBe("http://example.com");
  });
});
