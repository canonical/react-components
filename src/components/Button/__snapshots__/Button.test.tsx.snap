// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button  renders 1`] = `
<button
  className="p-button"
>
  Test content
</button>
`;

exports[`Button  renders as a link 1`] = `
<a
  className="p-button"
>
  Test content
</a>
`;
